### ISSUE TITLE
 MaskZeroCriterion and SequencerCriterion error

### ISSUE THREAD

**Issue by mzhai2** [Element-Research/rnn#130]
Using a MaskZeroCriterion and the SequencerCriterion works with floats but throws an error with cutorch. https://github.com/torch/cutorch/blob/master/lib/THC/generic/THCTensorCopy.c

`
/Users/Mike/torch/install/bin/luajit: ...ke/torch/install/share/lua/5.1/rnn/MaskZeroCriterion.lua:35: bad argument #1 to 'index' (sizes do not match at /tmp/luarocks_cutorch-scm-1-186/cutorch/lib/THC/generic/THCTensorCopy.c:47)
stack traceback:
	[C]: in function 'index'
	...ke/torch/install/share/lua/5.1/rnn/MaskZeroCriterion.lua:35: in function 'recursiveMask'
	...ke/torch/install/share/lua/5.1/rnn/MaskZeroCriterion.lua:75: in function 'forward'
	...e/torch/install/share/lua/5.1/rnn/SequencerCriterion.lua:39: in function 'forward'
	./train.lua:107: in function 'opfunc'
	/Users/Mike/torch/install/share/lua/5.1/optim/sgd.lua:44: in function 'optimMethod'
	./train.lua:131: in function 'train'
	main.lua:97: in main chunk
	[C]: in function 'dofile'
	...Mike/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x010570ebd0
`

**Comment by nicholas-leonard** [comment]
@mzhai2 I have added unit tests for testing cuda() for both MaskZeroCriterion and SequencerCriterion : 
https://github.com/Element-Research/rnn/commit/30d8f1cfee66a2745b6ecc39da02c594edcba211

Both tests pass at the moment. Could you test them on your machine? I they don't fail, try and modify them to reproduce your issue so I can reproduce on my end. Thank you.

**Comment by mzhai2** [comment]
For MaskZeroCriterion:updateOutput input is all zeros of size 4x53628 and target is all zeros of size 4. Sorry I am not too familiar with torch. It's failing for float tensors too.

/Users/Mike/torch/install/bin/luajit: ...ke/torch/install/share/lua/5.1/rnn/MaskZeroCriterion.lua:35: bad argument #2 to 'index' (Index is supposed to be a vector at /tmp/luarocks_torch-scm-1-873/torch7/lib/TH/generic/THTensorMath.c:115)
stack traceback:
	[C]: in function 'index'
	...ke/torch/install/share/lua/5.1/rnn/MaskZeroCriterion.lua:35: in function 'recursiveMask'
	...e/torch/install/share/lua/5.1/rnn/MaskZeroCriterion.lua:75: in function 'forward'
	./train.lua:114: in function 'opfunc'
	/Users/Mike/torch/install/share/lua/5.1/optim/sgd.lua:44: in function 'optimMethod'
	./train.lua:146: in function 'train'
	main.lua:97: in main chunk
	[C]: in function 'dofile'
	...Mike/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x010c4c9bd0
 
I'm guessing the problem is being caused by self.zeroMask that is a LongTensor with no dimensions when its passed to recursive mask, zeromask being [0, 0, 0, 0]

**Comment by nicholas-leonard** [comment]
@mzhai2 For ClassNLLCriterion don't use 0 for your dummy target tensors as it will crash ClassNLLCriterion. Remember, that criterion's forward and backward methods still get called. So instead use targets between 1 and 4. It is ok as their gradinputs and loss will be zeroed.

**Comment by mzhai2** [comment]
So for a language I should pad my input with 0s and each corresponding target should be between 1 and 4, right?

**Comment by nicholas-leonard** [comment]
So pad your input to the criterion. But initialize the targets to valid values (1 will always work). See how ClassNLLCriterion fails with zero targets : 
```lua
th> crit = nn.ClassNLLCriterion()
                                                                      [0.0026s]
th> crit:forward(torch.randn(3,4), torch.zeros(3))
/usr/local/share/lua/5.1/nn/THNN.lua:254: Assertion `cur_target >= 0 && cur_target < n_classes' failed.  at /tmp/luarocks_nn-scm-1-6917/nn/lib/THNN/generic/ClassNLLCriterion.c:41
stack traceback:
	[C]: in function 'v'
	/usr/local/share/lua/5.1/nn/THNN.lua:254: in function 'ClassNLLCriterion_updateOutput'
	/usr/local/share/lua/5.1/nn/ClassNLLCriterion.lua:41: in function 'forward'
	[string "crit:forward(torch.randn(3,4), torch.zeros(3))"]:1: in main chunk
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/trepl/init.lua:587: in function 'repl'
	/usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
	[C]: at 0x00405e60	
```

**Comment by mzhai2** [comment]
Ok, thanks for the clarification.
